File: packages/edraft/src/zip.ts
/**
 * Minimal ZIP reader — just enough of the format to open an Office Open XML
 * package (.docx). Zero dependencies: stored entries are sliced, deflated
 * entries inflate through the platform's native DecompressionStream.
 *
 * Doctrine: refuse loudly what we do not understand (ZIP64, encryption,
 * multi-disk, exotic compression). A clear error is a feature; a silent
 * misread is a bug.
 */

import { decodeUtf8, inflateRaw } from './platform.js';
import { crc32 } from './crc32.js';

/** The archive cannot be read as a ZIP — corrupt, encrypted, or exotic. */
export class ZipFormatError extends Error {
	constructor(message: string) {
		super(message);
		this.name = 'ZipFormatError';
	}
}

export interface ZipEntry {
	/** Full path inside the archive, e.g. 'word/document.xml'. */
	name: string;
	/** Uncompressed contents. */
	data: Uint8Array;
}

export interface ZipReadOptions {
	/** Refuse archives with more entries than this. Default 512. */
	maxEntries?: number;
	/** Refuse any single entry whose uncompressed size exceeds this. Default 64 MB. */
	maxEntryBytes?: number;
	/** Refuse when total uncompressed size exceeds this. Default 128 MB. */
	maxTotalBytes?: number;
}

export const DEFAULT_ZIP_MAX_ENTRIES: number = 512;
export const DEFAULT_ZIP_MAX_ENTRY_BYTES: number = 64 * 1024 * 1024;
export const DEFAULT_ZIP_MAX_TOTAL_BYTES: number = 128 * 1024 * 1024;

const EOCD_SIG = 0x06054b50;
const CENTRAL_SIG = 0x02014b50;
const LOCAL_SIG = 0x04034b50;

function u16(bytes: Uint8Array, at: number): number {
	return bytes[at]! | (bytes[at + 1]! << 8);
}

function u32(bytes: Uint8Array, at: number): number {
	return (bytes[at]! | (bytes[at + 1]! << 8) | (bytes[at + 2]! << 16) | (bytes[at + 3]! << 24)) >>> 0;
}

/**
 * Read every entry of a ZIP archive. Async because inflation is streamed.
 * Sizes are trusted from the CENTRAL directory (never the local header),
 * which is what makes data-descriptor archives safe to read.
 */
export async function readZipEntries(source: Uint8Array, options: ZipReadOptions = {}): Promise<ZipEntry[]> {
	const maxEntries = options.maxEntries ?? DEFAULT_ZIP_MAX_ENTRIES;
	const maxEntryBytes = options.maxEntryBytes ?? DEFAULT_ZIP_MAX_ENTRY_BYTES;
	const maxTotalBytes = options.maxTotalBytes ?? DEFAULT_ZIP_MAX_TOTAL_BYTES;
	if (source.length < 22) throw new ZipFormatError('not a ZIP archive (file too small)');

	/* the End Of Central Directory record lives in the last 64 KB + 22 bytes */
	let eocd = -1;
	const scanFrom = Math.max(0, source.length - 22 - 65535);
	for (let i = source.length - 22; i >= scanFrom; i--) {
		if (u32(source, i) === EOCD_SIG) {
			eocd = i;
			break;
		}
	}
	if (eocd < 0) throw new ZipFormatError('not a ZIP archive (end record missing)');

	if (u16(source, eocd + 4) !== 0 || u16(source, eocd + 6) !== 0) {
		throw new ZipFormatError('multi-disk archives are not supported');
	}
	const count = u16(source, eocd + 10);
	const centralOffset = u32(source, eocd + 16);
	if (count === 0xffff || centralOffset === 0xffffffff) {
		throw new ZipFormatError('ZIP64 archives are not supported');
	}
	if (count > maxEntries) throw new ZipFormatError(`archive holds ${count} entries — over the ${maxEntries} limit`);

	const entries: ZipEntry[] = [];
	let totalBytes = 0;
	let at = centralOffset;
	for (let n = 0; n < count; n++) {
		if (u32(source, at) !== CENTRAL_SIG) throw new ZipFormatError('corrupt central directory');
		const flags = u16(source, at + 8);
		const method = u16(source, at + 10);
		const recordedCrc = u32(source, at + 16);
		const packedSize = u32(source, at + 20);
		const rawSize = u32(source, at + 24);
		const nameLength = u16(source, at + 28);
		const extraLength = u16(source, at + 30);
		const commentLength = u16(source, at + 32);
		const localOffset = u32(source, at + 42);
		/* office document names are ASCII; UTF-8 bytes survive decode either way */
		const name = decodeUtf8(source.subarray(at + 46, at + 46 + nameLength));

		if ((flags & 0x1) !== 0) throw new ZipFormatError(`entry "${name}" is encrypted — not supported`);
		if (rawSize > maxEntryBytes) {
			throw new ZipFormatError(`entry "${name}" is ${rawSize} bytes — over the ${maxEntryBytes} limit`);
		}
		if (u32(source, localOffset) !== LOCAL_SIG) {
			throw new ZipFormatError(`corrupt local header for "${name}"`);
		}
		const localNameLength = u16(source, localOffset + 26);
		const localExtraLength = u16(source, localOffset + 28);
		const dataStart = localOffset + 30 + localNameLength + localExtraLength;
		const packed = source.subarray(dataStart, dataStart + packedSize);

		let data: Uint8Array;
		if (method === 0) {
			data = packed.slice();
		} else if (method === 8) {
			try {
				data = await inflateRaw(packed, rawSize);
			} catch (error) {
				throw new ZipFormatError(
					`entry "${name}" could not be inflated: ${error instanceof Error ? error.message : 'unknown failure'}`
				);
			}
		} else {
			throw new ZipFormatError(`entry "${name}" uses compression method ${method} — not supported`);
		}
		if (crc32(data) !== recordedCrc) {
			throw new ZipFormatError(`entry "${name}" failed its integrity check (CRC mismatch)`);
		}
		totalBytes += data.length;
		if (totalBytes > maxTotalBytes) {
			throw new ZipFormatError(`archive expands past the ${maxTotalBytes}-byte limit`);
		}
		entries.push({ name, data });
		at += 46 + nameLength + extraLength + commentLength;
	}
	return entries;
}

/** One entry as the tolerant reader found it (docs/RFC-DRAFT-FORMAT.md §7.2). */
export interface ZipTolerantEntry {
	/** The name as the archive spells it, decoded as UTF-8. */
	name: string;
	/** The name's bytes, so a caller can tell whether they were valid UTF-8. */
	nameBytes: Uint8Array;
	/** The entry's contents: uncompressed when it could be read, otherwise
		whatever bytes the archive holds for it (see `error`). */
	data: Uint8Array;
	/** Why the entry could not be read whole — absent when it was. */
	error?: string;
}

export interface ZipTolerantResult {
	entries: ZipTolerantEntry[];
	/** Where the entries came from: the central directory, or — when it is
		missing or damaged, as in a truncated file — the local headers. */
	directory: 'central' | 'local';
}

export interface ZipTolerantReadOptions extends ZipReadOptions {
	/** Largest uncompressed-to-compressed ratio an entry may declare. Default 200. */
	maxRatio?: number;
}

export const DEFAULT_ZIP_MAX_RATIO: number = 200;

/**
 * Read every entry that can be read, and keep the bytes of every entry that
 * cannot — the .draft reader's recovery ladder stands on this. Unlike
 * `readZipEntries`, one damaged entry does not refuse the archive: it comes
 * back with an `error` and its bytes. A declared size is never trusted to
 * allocate: an entry that declares more than the limits allow is not
 * expanded, and says so. Only an archive that is not a ZIP at all, or holds
 * more entries than allowed, is refused.
 */
export async function readZipEntriesTolerant(
	source: Uint8Array,
	options: ZipTolerantReadOptions = {}
): Promise<ZipTolerantResult> {
	const maxEntries = options.maxEntries ?? DEFAULT_ZIP_MAX_ENTRIES;
	const central = centralRecords(source);
	if (central !== null && central.length > maxEntries) {
		throw new ZipFormatError(`archive holds ${central.length} entries — over the ${maxEntries} limit`);
	}
	const records = central ?? localRecords(source, maxEntries);
	if (records === null) throw new ZipFormatError('not a ZIP archive');
	const entries: ZipTolerantEntry[] = [];
	let total = 0;
	for (const record of records) {
		const entry = await readRecord(source, record, options, total);
		if (entry.error === undefined) total += entry.data.length;
		entries.push(entry);
	}
	return { entries, directory: central === null ? 'local' : 'central' };
}

interface ZipRecord {
	nameBytes: Uint8Array;
	flags: number;
	method: number;
	crc: number;
	packedSize: number;
	rawSize: number;
	dataStart: number;
	/** Set when the record itself shows the entry cannot be read. */
	problem?: string;
}

/** The central directory's records, or null when it is missing or damaged. */
function centralRecords(source: Uint8Array): ZipRecord[] | null {
	if (source.length < 22) return null;
	let eocd = -1;
	const scanFrom = Math.max(0, source.length - 22 - 65535);
	for (let i = source.length - 22; i >= scanFrom; i--) {
		if (u32(source, i) === EOCD_SIG) {
			eocd = i;
			break;
		}
	}
	if (eocd < 0) return null;
	if (u16(source, eocd + 4) !== 0 || u16(source, eocd + 6) !== 0) return null;
	const count = u16(source, eocd + 10);
	const centralOffset = u32(source, eocd + 16);
	if (count === 0xffff || centralOffset === 0xffffffff) return null;
	const records: ZipRecord[] = [];
	let at = centralOffset;
	for (let n = 0; n < count; n++) {
		if (at + 46 > source.length || u32(source, at) !== CENTRAL_SIG) return null;
		const nameLength = u16(source, at + 28);
		const extraLength = u16(source, at + 30);
		const commentLength = u16(source, at + 32);
		const localOffset = u32(source, at + 42);
		if (at + 46 + nameLength > source.length) return null;
		const record: ZipRecord = {
			nameBytes: source.slice(at + 46, at + 46 + nameLength),
			flags: u16(source, at + 8),
			method: u16(source, at + 10),
			crc: u32(source, at + 16),
			packedSize: u32(source, at + 20),
			rawSize: u32(source, at + 24),
			dataStart: 0
		};
		if (localOffset + 30 > source.length || u32(source, localOffset) !== LOCAL_SIG) {
			record.problem = 'its local header is damaged';
		} else {
			record.dataStart = localOffset + 30 + u16(source, localOffset + 26) + u16(source, localOffset + 28);
		}
		records.push(record);
		at += 46 + nameLength + extraLength + commentLength;
	}
	return records;
}

/**
 * Records found by walking the local headers from the start — the way to
 * read a file whose end is missing. Stops at the first header that is not
 * whole. Null when the file does not begin with one.
 */
function localRecords(source: Uint8Array, maxEntries: number): ZipRecord[] | null {
	if (source.length < 30 || u32(source, 0) !== LOCAL_SIG) return null;
	const records: ZipRecord[] = [];
	let at = 0;
	while (at + 30 <= source.length && u32(source, at) === LOCAL_SIG) {
		if (records.length === maxEntries) {
			throw new ZipFormatError(`archive holds more than ${maxEntries} entries — over the limit`);
		}
		const flags = u16(source, at + 6);
		const nameLength = u16(source, at + 26);
		const extraLength = u16(source, at + 28);
		const dataStart = at + 30 + nameLength + extraLength;
		if (dataStart > source.length) break;
		const record: ZipRecord = {
			nameBytes: source.slice(at + 30, at + 30 + nameLength),
			flags,
			method: u16(source, at + 8),
			crc: u32(source, at + 14),
			packedSize: u32(source, at + 18),
			rawSize: u32(source, at + 22),
			dataStart
		};
		records.push(record);
		/* A data descriptor leaves the sizes out of the local header, so the
           next header cannot be found: stop at this entry. */
		if ((flags & 0x8) !== 0) {
			record.problem = 'its sizes are only in a data descriptor';
			break;
		}
		at = dataStart + record.packedSize;
	}
	return records;
}

async function readRecord(
	source: Uint8Array,
	record: ZipRecord,
	options: ZipTolerantReadOptions,
	totalSoFar: number
): Promise<ZipTolerantEntry> {
	const maxEntryBytes = options.maxEntryBytes ?? DEFAULT_ZIP_MAX_ENTRY_BYTES;
	const maxTotalBytes = options.maxTotalBytes ?? DEFAULT_ZIP_MAX_TOTAL_BYTES;
	const maxRatio = options.maxRatio ?? DEFAULT_ZIP_MAX_RATIO;
	const name = decodeUtf8(record.nameBytes);
	const base = { name, nameBytes: record.nameBytes };
	const end = Math.min(source.length, record.dataStart + record.packedSize);
	const packed = record.problem === undefined ? source.slice(record.dataStart, end) : new Uint8Array(0);
	const damaged = (error: string): ZipTolerantEntry => ({ ...base, data: packed, error });

	if (record.problem !== undefined) return damaged(record.problem);
	if ((record.flags & 0x1) !== 0) return damaged('it is encrypted');
	if (record.dataStart + record.packedSize > source.length) return damaged('it is cut short');
	if (record.rawSize > maxEntryBytes) return damaged(`it declares ${record.rawSize} bytes, over the ${maxEntryBytes} limit`);
	if (totalSoFar + record.rawSize > maxTotalBytes) return damaged(`the archive would expand past the ${maxTotalBytes}-byte limit`);

	let data: Uint8Array;
	if (record.method === 0) {
		if (record.rawSize !== record.packedSize) return damaged('its sizes disagree');
		data = packed;
	} else if (record.method === 8) {
		if (record.rawSize > Math.max(1, record.packedSize) * maxRatio) {
			return damaged(`it declares a compression ratio over ${maxRatio}:1`);
		}
		try {
			data = await inflateRaw(packed, record.rawSize);
		} catch {
			return damaged('it could not be inflated');
		}
		if (data.length !== record.rawSize) return damaged('it inflated to a different size than it declares');
	} else {
		return damaged(`it uses compression method ${record.method}`);
	}
	if (crc32(data) !== record.crc) return { ...base, data, error: 'it failed its CRC-32 check' };
	return { ...base, data };
}

/** The entry with this exact archive path, or undefined. */
export function findEntry(entries: ZipEntry[], name: string): ZipEntry | undefined {
	return entries.find((entry) => entry.name === name);
}

File: apple/eDraftEngine/Sources/EDraftEngine/Export/Zip.swift
import Compression
import Foundation

/// ZIP, as much as the .draft file needs (docs/RFC-DRAFT-FORMAT.md §4):
/// a writer of stored entries, and a tolerant reader that keeps the bytes of
/// every entry it cannot read instead of refusing the archive. Mirrors the
/// TypeScript engine's `zipwrite.ts` and `readZipEntriesTolerant` in
/// `zip.ts`; `Fixtures/draft.json` pins both, byte for byte.
public enum Zip {

    public struct Entry: Equatable, Sendable {
        public var name: String
        public var data: [UInt8]
        public init(name: String, data: [UInt8]) {
            self.name = name
            self.data = data
        }
    }

    public struct TolerantEntry: Equatable, Sendable {
        /// The name as the archive spells it, decoded as UTF-8.
        public var name: String
        public var nameBytes: [UInt8]
        /// Uncompressed when the entry could be read; otherwise whatever
        /// bytes the archive holds for it.
        public var data: [UInt8]
        /// Why the entry could not be read whole — nil when it was.
        public var error: String?
    }

    public struct TolerantResult: Equatable, Sendable {
        public var entries: [TolerantEntry]
        /// "central", or "local" when the central directory is missing or
        /// damaged and the entries came from the local headers.
        public var directory: String
    }

    public struct FormatError: Error, Equatable, Sendable {
        public let message: String
    }

    public static let maxEntries = 512
    public static let maxEntryBytes = 64 * 1024 * 1024
    public static let maxTotalBytes = 128 * 1024 * 1024
    public static let maxRatio = 200

    private static let localSig: UInt32 = 0x0403_4B50
    private static let centralSig: UInt32 = 0x0201_4B50
    private static let endSig: UInt32 = 0x0605_4B50

    // MARK: - Writing

    /// A single-disk ZIP of stored entries. The defaults are the TypeScript
    /// writer's historic bytes; the .draft writer asks for 1980-01-01 and the
    /// UTF-8 flag (§4.1, §4.4).
    public static func writeStored(
        _ entries: [Entry], dosDate: UInt16 = 0, dosTime: UInt16 = 0, utf8Names: Bool = false
    ) -> [UInt8] {
        var out: [UInt8] = []
        var central: [UInt8] = []
        for entry in entries {
            let name = Array(entry.name.utf8)
            let crc = CRC32.checksum(entry.data)
            let flags: UInt16 = utf8Names && name.contains(where: { $0 > 0x7F }) ? 0x0800 : 0
            let offset = UInt32(out.count)
            out += le32(localSig) + le16(20) + le16(flags) + le16(0) + le16(dosTime) + le16(dosDate)
            out += le32(crc) + le32(UInt32(entry.data.count)) + le32(UInt32(entry.data.count))
            out += le16(UInt16(name.count)) + le16(0) + name + entry.data
            central += le32(centralSig) + le16(20) + le16(20) + le16(flags) + le16(0) + le16(dosTime) + le16(dosDate)
            central += le32(crc) + le32(UInt32(entry.data.count)) + le32(UInt32(entry.data.count))
            central += le16(UInt16(name.count)) + le16(0) + le16(0) + le16(0) + le16(0) + le32(0) + le32(offset) + name
        }
        let centralOffset = UInt32(out.count)
        out += central
        out += le32(endSig) + le16(0) + le16(0) + le16(UInt16(entries.count)) + le16(UInt16(entries.count))
        out += le32(UInt32(central.count)) + le32(centralOffset) + le16(0)
        return out
    }

    // MARK: - Reading

    /// Every entry that can be read, and the bytes of every entry that
    /// cannot. Only an archive that is not a ZIP, or holds more entries than
    /// allowed, is refused. A declared size is never trusted to allocate.
    public static func readTolerant(_ source: [UInt8]) throws -> TolerantResult {
        let central = centralRecords(source)
        if let central, central.count > maxEntries {
            throw FormatError(message: "archive holds \(central.count) entries — over the \(maxEntries) limit")
        }
        guard let records = try central ?? localRecords(source) else {
            throw FormatError(message: "not a ZIP archive")
        }
        var entries: [TolerantEntry] = []
        var total = 0
        for record in records {
            let entry = read(source, record, totalSoFar: total)
            if entry.error == nil { total += entry.data.count }
            entries.append(entry)
        }
        return TolerantResult(entries: entries, directory: central == nil ? "local" : "central")
    }

    private struct Record {
        var nameBytes: [UInt8]
        var flags: UInt16
        var method: UInt16
        var crc: UInt32
        var packedSize: Int
        var rawSize: Int
        var dataStart: Int
        var problem: String?
    }

    private static func centralRecords(_ source: [UInt8]) -> [Record]? {
        guard source.count >= 22 else { return nil }
        var eocd = -1
        let scanFrom = max(0, source.count - 22 - 65535)
        var i = source.count - 22
        while i >= scanFrom {
            if u32(source, i) == endSig { eocd = i; break }
            i -= 1
        }
        guard eocd >= 0, u16(source, eocd + 4) == 0, u16(source, eocd + 6) == 0 else { return nil }
        let count = Int(u16(source, eocd + 10))
        let centralOffset = u32(source, eocd + 16)
        guard count != 0xFFFF, centralOffset != 0xFFFF_FFFF else { return nil }
        var records: [Record] = []
        var at = Int(centralOffset)
        for _ in 0..<count {
            guard at + 46 <= source.count, u32(source, at) == centralSig else { return nil }
            let nameLength = Int(u16(source, at + 28))
            let extraLength = Int(u16(source, at + 30))
            let commentLength = Int(u16(source, at + 32))
            let localOffset = Int(u32(source, at + 42))
            guard at + 46 + nameLength <= source.count else { return nil }
            var record = Record(
                nameBytes: Array(source[(at + 46)..<(at + 46 + nameLength)]),
                flags: u16(source, at + 8),
                method: u16(source, at + 10),
                crc: u32(source, at + 16),
                packedSize: Int(u32(source, at + 20)),
                rawSize: Int(u32(source, at + 24)),
                dataStart: 0
            )
            if localOffset + 30 > source.count || u32(source, localOffset) != localSig {
                record.problem = "its local header is damaged"
            } else {
                record.dataStart = localOffset + 30 + Int(u16(source, localOffset + 26)) + Int(u16(source, localOffset + 28))
            }
            records.append(record)
            at += 46 + nameLength + extraLength + commentLength
        }
        return records
    }

    private static func localRecords(_ source: [UInt8]) throws -> [Record]? {
        guard source.count >= 30, u32(source, 0) == localSig else { return nil }
        var records: [Record] = []
        var at = 0
        while at + 30 <= source.count, u32(source, at) == localSig {
            if records.count == maxEntries {
                throw FormatError(message: "archive holds more than \(maxEntries) entries — over the limit")
            }
            let flags = u16(source, at + 6)
            let nameLength = Int(u16(source, at + 26))
            let extraLength = Int(u16(source, at + 28))
            let dataStart = at + 30 + nameLength + extraLength
            if dataStart > source.count { break }
            var record = Record(
                nameBytes: Array(source[(at + 30)..<(at + 30 + nameLength)]),
                flags: flags,
                method: u16(source, at + 8),
                crc: u32(source, at + 14),
                packedSize: Int(u32(source, at + 18)),
                rawSize: Int(u32(source, at + 22)),
                dataStart: dataStart
            )
            if flags & 0x8 != 0 {
                record.problem = "its sizes are only in a data descriptor"
                records.append(record)
                break
            }
            records.append(record)
            at = dataStart + record.packedSize
        }
        return records
    }

    private static func read(_ source: [UInt8], _ record: Record, totalSoFar: Int) -> TolerantEntry {
        let name = String(decoding: record.nameBytes, as: UTF8.self)
        let end = min(source.count, record.dataStart + record.packedSize)
        let packed = record.problem == nil && record.dataStart <= end ? Array(source[record.dataStart..<end]) : []
        func damaged(_ error: String) -> TolerantEntry {
            TolerantEntry(name: name, nameBytes: record.nameBytes, data: packed, error: error)
        }
        if let problem = record.problem { return damaged(problem) }
        if record.flags & 0x1 != 0 { return damaged("it is encrypted") }
        if record.dataStart + record.packedSize > source.count { return damaged("it is cut short") }
        if record.rawSize > maxEntryBytes {
            return damaged("it declares \(record.rawSize) bytes, over the \(maxEntryBytes) limit")
        }
        if totalSoFar + record.rawSize > maxTotalBytes {
            return damaged("the archive would expand past the \(maxTotalBytes)-byte limit")
        }
        var data: [UInt8]
        switch record.method {
        case 0:
            if record.rawSize != record.packedSize { return damaged("its sizes disagree") }
            data = packed
        case 8:
            if record.rawSize > max(1, record.packedSize) * maxRatio {
                return damaged("it declares a compression ratio over \(maxRatio):1")
            }
            guard let inflated = inflate(packed, declared: record.rawSize) else {
                return damaged("it could not be inflated")
            }
            if inflated.count != record.rawSize { return damaged("it inflated to a different size than it declares") }
            data = inflated
        default:
            return damaged("it uses compression method \(record.method)")
        }
        if CRC32.checksum(data) != record.crc {
            return TolerantEntry(name: name, nameBytes: record.nameBytes, data: data, error: "it failed its CRC-32 check")
        }
        return TolerantEntry(name: name, nameBytes: record.nameBytes, data: data, error: nil)
    }

    /// Raw DEFLATE into a buffer one byte larger than declared, so an entry
    /// that inflates past its record is caught at the record, never at the
    /// memory limit. Nil when the stream is not DEFLATE.
    private static func inflate(_ packed: [UInt8], declared: Int) -> [UInt8]? {
        if packed.isEmpty { return declared == 0 ? [] : nil }
        let capacity = declared + 1
        var out = [UInt8](repeating: 0, count: capacity)
        let written = out.withUnsafeMutableBufferPointer { dst in
            packed.withUnsafeBufferPointer { src in
                compression_decode_buffer(dst.baseAddress!, capacity, src.baseAddress!, packed.count, nil, COMPRESSION_ZLIB)
            }
        }
        if written == 0 && declared > 0 { return nil }
        return Array(out[0..<written])
    }

    // MARK: - Bytes

    private static func u16(_ b: [UInt8], _ at: Int) -> UInt16 {
        guard at + 2 <= b.count, at >= 0 else { return 0 }
        return UInt16(b[at]) | UInt16(b[at + 1]) << 8
    }

    private static func u32(_ b: [UInt8], _ at: Int) -> UInt32 {
        guard at + 4 <= b.count, at >= 0 else { return 0 }
        return UInt32(b[at]) | UInt32(b[at + 1]) << 8 | UInt32(b[at + 2]) << 16 | UInt32(b[at + 3]) << 24
    }

    private static func le16(_ v: UInt16) -> [UInt8] { [UInt8(v & 0xFF), UInt8(v >> 8)] }

    private static func le32(_ v: UInt32) -> [UInt8] {
        [UInt8(v & 0xFF), UInt8((v >> 8) & 0xFF), UInt8((v >> 16) & 0xFF), UInt8(v >> 24)]
    }
}
